adding glue and paste functionality

choldgraf opened the pull request:
This adds a prototype functionality for "glue and paste" with MyST-NB. It closes #4 

It doesn't yet have tests because I want feedback on the API and functionality first!

You glue things into a notebook's metadata like this:

```python
from myst_nb import glue
glue("your-key", an_object)
```

And it will run IPython's display on the object, then store the mimebundle
at the key you specify.

When the notebooks are parsed with `MyST-NB`, it builds up a registry of all
the keys across all notebooks, so that you can then refer to them in the following ways:

You can paste it in markdown with a directive like this:

````
```{paste} your-key
```
````

Or you can add it in-line like this:

```
{paste}`your-key`
```
optionally:

```
{paste}`your-key:format-string`
```

Here's a docs page demo-ing how this works: https://141-241268229-gh.circle-artifacts.com/0/html/use/glue.html

I think there are still some other things we'd like to add, like adding captions to figures somehow etc, but I'm not sure whether that should be done with `glue` functions, or if it should be handled by the `MimeOutputBundle` transform.

Would love to hear thoughts! cc @chrisjsewell @mmcky @jstac and @AakashGfude (that's the group that first discussed this feature)

chrisjsewell commented on myst_nb/glue.py:
```diff
@@ -0,0 +1,173 @@
+                    bundle = {this_key: val for key, val in bundle.items()}
+                    outputs.append(bundle)
+    if len(outputs) == 0:
+        print(f"Warning: did not find key {this_key} in notebook {path_ntbk}")
```
```suggestion
        SPHINX_LOGGER.warning(f"Did not find key {this_key} in notebook {path_ntbk}")
```

chrisjsewell replied:
I think you accidentally reverted this in your force push

chrisjsewell commented on myst_nb/glue.py:
```diff
@@ -0,0 +1,173 @@
+    if len(outputs) > 1:
+        print(
+            f"Warning: multiple variables found for key: {this_key}. Returning first value."
+        )
```
```suggestion
        SPHINX_LOGGER.warning(
            f"Multiple variables found for key: {this_key}. Returning first value."
        )
```

chrisjsewell replied:
ditto

chrisjsewell commented on myst_nb/parser.py:
```diff
@@ -124,14 +125,40 @@ def parse(self, inputstring, document):
             pass
 
-        # Write the notebook's output to disk
+        # Remove all the mime prefixes from "glue" step.
```
maybe add a configuration option to turn this functionality on/off

choldgraf replied:
This is needed in order to ensure that the files actually get written to disk. The jupyter-sphinx writer step only writes image files that it has prefixes for. So if we have the `nteract/application` stuff in there it'll skip them.

Another option would be to try and upstream some logic to `jupyter-sphinx` that'll check for a prefix...

chrisjsewell replied:
I meant to turn off the entire glue functionality, but obviously anythingupstreamed tojupyter-sphinx is probably good

choldgraf replied:
@chrisjsewell ah that's a good point, agreed

choldgraf replied:
@akhmerov would you be +1 on an addition to the "write_outputs" function that would allow us to add a "prefix" to any outputs that it would also search for? That way `image/png` would work as well as `myprefix/image/png`.

chrisjsewell commented:
In general, I think it could be good; as long as the implementation doesn't become too complex/error prone

chrisjsewell commented:
> I think there are still some other things we'd like to add, like adding captions to figures somehow 

We definitely need to get round to this sooner rather than later. Adding captions, referencable labels and other general formatting (size, alignment, etc) to code outputs is a major aspect of what we want to achieve IMO

choldgraf commented:
@chrisjsewell well, we have a directive that we *could* do other things with. Perhaps we can pass some more metadata through the `OutputBundle` nodes, and then we can add functionality for things like captions in figures etc in the post-transform step?

chrisjsewell commented:
> @chrisjsewell re: captions,references, etc we do have a directive that we _could_ do other things with. Perhaps we can pass some more metadata through the `OutputBundle` nodes, and then we can add functionality for things like captions in figures etc in the post-transform step?

Yeh definitely, something like that 👍

chrisjsewell commented:
With the paste role/directive we also have the opportunity to exert a lot of control over the output. I could envisage creating a [sphinx domain](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html), e.g.

````
```{paste:literal} key1
```

```{paste:figure} key2
:name: label

A caption...
```

```{paste:subfig}
- key3
- key4
```
````

chrisjsewell commented:
> With the paste role/directive we also have the opportunity to exert a lot of control over the output. I could envisage creating a sphinx domain, e.g.

If we go this route, although I like the word `paste`, it could be a bit long, so perhaps `nb` would be better?

```rst
:nb:text:`your-key:format-string`
```

chrisjsewell commented on myst_nb/glue.py:
```diff
@@ -0,0 +1,173 @@
+mime_prefix = "application/papermill.record/"
+
+
+def glue(name, variable):
```
For things like figures, we probably want to show them in the notebook, at the same time as 'glueing' them, e.g. `glue("key", fig, display=True)`.

a) This would probably be used at the same time as tagging the cell with `remove_output`
b) In this use case we probably wouldn't want to prefix the mime types, because then we would have to store duplicate prefixed/non-prefixed outputs, so we would want to indicate whether the mime types had been prefixed, e.g.

`metadata["scrapbook"] = dict(name=name, mime_prefix=None)`

choldgraf replied:
yeah - my thinking was that people probably have already shown the figure, since it gets shown by default when you create the matplotlib object. At least when I've used scrapbook, I *always* end up having two figures plotted side-by-side, which is why I didn't feel the need to implement this in the current pass (since the act of "gluing" a plot is orthogonal to the act of "displaying" a plot). Do you feel strongly that this needs to be possible?

chrisjsewell replied:
Yes matplotlib is a bit of an outlier in this respect, in that basically it doesn't use the `IPython` mechanics properly. All other packages, like pandas, (correctly) don't have this 'auto-display' behaviour, and so this is where you would use it.
(you can also turn off matplotlib's auto show I think: https://stackoverflow.com/questions/18717877/prevent-plot-from-showing-in-jupyter-notebook)

choldgraf commented:
> :nb:text:`your-key:format-string`

Interesting idea - I do like the idea of giving more specific commands when calling the role/directive (e.g. `{nb:text}` and `{nb:fig}`), though it's also a bit more complex to remember...and probably outside of my sphinx knowledge. I am happy to merge this as a super-alpha module, if you want to hack it around and improve the API. What do you think?

chrisjsewell commented:
> I am happy to merge this as a super-alpha module to get a feel for how the API works

Yeh, address this:
https://github.com/ExecutableBookProject/MyST-NB/pull/66#discussion_r392030315,
and fix the tests, then probably we can merge and add the role/directive domain idea as an `enhancement` issue

choldgraf commented:
sounds good - I'll add back in the display stuff and merge it in after some tests that pass

chrisjsewell commented on myst_nb/__init__.py:
```diff
@@ -35,6 +36,11 @@ def update_togglebutton_classes(app, config):
 
+def init_glue_cache(app):
+    if not hasattr(app.env, "glue_data"):
+        app.env.glue_data = {}
```
It would be nice to have 'external' access to this dict, to see what keys we have stored. Maybe we can write a small function in myst-nb, that `ExecutableBookProject/cli` could utilise (outside of sphinx) to show the currently stored keys:

```python
from pathlib import Path
import pickle

def get_glue_data(doctreedir):
    path = Path(doctreedir, "environment.pickle")
    env = pickle.loads(path.read_bytes())
    return env.glue_data
```

chrisjsewell replied:
Or alternatively you store this as a separate JSON file in the `doctreesdir`.

chrisjsewell commented on myst_nb/glue.py:
```diff
@@ -0,0 +1,177 @@
+                formatting = None
+
+            if key not in glue_data:
+                raise KeyError(
```
I think this should use the sphinx logging mechanics, to also report the `paste_node` source file and line number

chrisjsewell replied:
Also, should we be raising exceptions, that block the whole build, or just reporting warnings/errors

chrisjsewell commented on myst_nb/glue.py:
```diff
@@ -0,0 +1,177 @@
+    option_spec = {"id": directives.unchanged}
+
+    def run(self):
+        return [PasteNode(self.arguments[0], "directive")]
```
I think you need to add the source path / line number to the node here, for later use in reporting (have a look at other sphinx directives)

chrisjsewell commented on myst_nb/glue.py:
```diff
@@ -0,0 +1,177 @@
+                            pass
+                    out_node = nodes.inline(text, text, classes=["pasted-text"])
+                else:
+                    SPHINX_LOGGER.warn(
```
ditto reporting source file / line number

chrisjsewell commented on myst_nb/glue.py:
```diff
@@ -0,0 +1,177 @@
+                out_node += CellInputNode()
+                out_node += output_node
+            else:
+                raise ValueError(
```
ditto reporting source file / line number

choldgraf commented:
I think I just addressed all your comments in the latest commit - LMK if that looks correct to you. I'm still not sure how to handle line numbers with the ipynb files, but in the meantime I'm just removing the line number from warnings with an `.ipynb` extension, and we can figure that out later.

We still need to add tests for this. At the least:

* Ensure `glue` works as expected, puts in the correct mime information etc
* Ensure the role + directive grab the correct glued items and insert them in the doctree
* Ensure the glue cache is correct after a build

Anything else?

choldgraf commented:
(also, I am not really sure how to test this...it seems that our test infrastructure is just running the parser + transform on a single file, whereas this requires us to test something that collects across multiple files in a sphinx site)

chrisjsewell commented:
I’m on it...

choldgraf commented:
@chrisjsewell I just pushed another commit that adds some very basic tests that don't quite work. It uses teh regression framework you have, but it doesn't yet do multi-page testing, and also I can't figure out how to register the new roles+directives in sphinx.

Happy for you to clobber those changes if you like :-)

chrisjsewell commented on myst_nb/glue.py:
```diff
@@ -0,0 +1,234 @@
+class PasteNodesToDocutils(SphinxTransform):
+    """Use the builder context to transform a CellOutputNode into Sphinx nodes."""
+
+    default_priority = 700
```
We should add a comment of justification for why 700 was chosen, and ditto for `CellOutputsToNodes`. Also should one be a higher priority than the other?

choldgraf replied:
Ummmm...that makes a very strong assumption that I have logical reasoning behind my decisions...

But I did a [quick wikipedia search](https://en.wikipedia.org/wiki/700_(number)) and this is why I've decided that the number is 700:

> 700 (seven hundred) is the natural number following 699 and preceding 701. It is the sum of four consecutive primes (167 + 173 + 179 + 181).

chrisjsewell replied:
😆

chrisjsewell replied:
Yeh `PasteNodesToDocutils` needs to be run before `CellOutputsToNodes` (see 
72a5595)

chrisjsewell commented:
tests are done

chrisjsewell commented:
We could add a full sphinx integration test (like in myst-parser), but I'll save that for a later date

choldgraf commented:
You are amazing, much appreciated 🙏

choldgraf commented:
Are you +1 on merging this, and opening an issue to discuss the role / directive names etc?

chrisjsewell commented:
> Are you +1 on merging this, and opening an issue to discuss the role / directive names etc?

Once you resolve https://github.com/ExecutableBookProject/MyST-NB/pull/66#discussion_r392544779, yes 😁

choldgraf commented:
oops - I just saw this...thanks for changing the priority

chrisjsewell commented:
No problem, I jumped in to merge it before you, so that I could add a proper commit message 😉
